---
title: "## INSERT PROJECT TITLE ## - neo-antigen screen"
author: "Lorenzo Fanchi"
date: "`r format(Sys.Date(), '%d-%b-%Y')`"
output: 
  html_document: 
    theme: cerulean
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
options(knitr.table.format = 'html')

setwd('~/projects/## PROJECT DIR ##')

source('runConfig.R')
source('helperFunctions.R')
```

<link href="style.css" rel="stylesheet" type="text/css" />

<br>

#### Introduction

This document contains the order and set-up lists of the epitopes for the ## PROJECT TITLE ## project.

<br>

#### Methods summary

Samples were sequenced by GCF at the Netherlands Cancer Institute, data can be found at:  

* [GCF DNA repository link](### INSERT LINK HERE ###)  
* [GCF RNA repository link](### INSERT LINK HERE ###)  

The GCF performed the following:  

1. DNA (exome) and RNA sequencing  
2. variant calling using GATK MuTect2 & HaplotypeCaller for somatic and germline calls, respectively  
3. generation of VCF files containing somatic and germline calls  

I further processed the resulting VCF files and RNAseq data by:  

1. annotating identified variants with 'expression of mutant allele' data  
2. generating variant contexts for all expressed variants (or variants without expression data) using the varcontext tool (branch: #####, commit: #####)  
3. aligning RNAseq data using STAR & salmon, generating TPM values  
4. annotating varcontext output with gene-level expression data  

Point 1 was done by performing pileups at the SNV genomic loci in the aligned RNAseq BAM files. For a given SNV locus, a minimum of 5 or more reads for either the germline or the mutant base were required for further analysis. In case this requirement was met, expression of the mutant allele was deemed present if > 1 read containing the mutant base was found. In case minimum read requirement was not met, the SNV was annotated with NA.  

Epitopes were predicted using in-house epitope prediction pipeline, using random forest prediction for probability of presentation (branch: #####, commit: #####)  

<br>

* Transcript effects of called variants were determined by our in-house 'varcontext' tool with the following parameters:  

```{r varcontext_info}
varcontext_dt = data.table(parameter = c('Ensembl build',
                                         'Genome assembly',
                                         'Mutant allele expression filter',
                                         'Canonical transcripts only'),
                           value = c(runOptions$varcontext$ensembl_build,
                                     runOptions$varcontext$assembly_build,
                                     as.character(runOptions$varcontext$mutant_allele_expression_filter),
                                     as.character(runOptions$varcontext$canonical_only)))

knitr::kable(varcontext_dt, align = 'c') %>%
  kable_styling(bootstrap_options = c("striped", "hover"), full_width = F, position = 'left') %>%
  column_spec(column = 1, bold = T) %>%
  column_spec(column = 2, width = '8cm')
```

<br>

* The following filtering parameters were used for epitope prediction:  

```{r epitope_pred_info}
prediction_dt = data.table(parameter = c('netMHCpan rank cutoff',
                                         'netChop processing cutoff',
                                         'Model cutoff',
                                         'RNA expression cutoff',
                                         'Random forest model',
                                         'Selfsim filter',
                                         'Selfsim mode',
                                         'Selflist'),
                           value = c(runOptions$neolution$rank_cutoff,
                                     runOptions$neolution$processing_cutoff,
                                     paste('>=', runOptions$neolution$model_cutoff),
                                     paste('>', runOptions$neolution$expression_cutoff),
                                     runOptions$neolution$random_forest_model,
                                     runOptions$neolution$selfsim_filter,
                                     runOptions$neolution$selfsim_filter_mode,
                                     runOptions$neolution$selflist))

knitr::kable(prediction_dt, align = 'c') %>%
  kable_styling(bootstrap_options = c('striped', 'hover'), full_width = F, position = 'left') %>%
  column_spec(column = 1, bold = T) %>%
  column_spec(column = 2, width = '8cm')
```

<br>

#### Sample information
__Selected alleles included in output are highlighted in green, excluded alleles in red__

```{r variant_snpEff_info}
snpeff_output = list.files(path = '4_snpEff',
                           pattern = 'genes[.]txt$',
                           full.names = T)
snpeff_output = snpeff_output[!grepl(pattern = 'snps', x = snpeff_output, fixed = TRUE)]

sample_info = fread('sample_info.tsv')
sample_info = sample_info[naturalorder(sample_info$patient_id)]

effects_table = rbindlist(lapply(snpeff_output,
                                 function(path) {
                                   data = fread(path, skip = 1)
                                   
                                   sample_prefix = gsub(pattern = regexPatterns$seqdata_prefix,
                                                        replacement = '\\1',
                                                        x = basename(path),
                                                        perl = T)
                                   
                                   if (sample_prefix %nin% sample_info$dna_data_prefix) {
                                     message(sample_prefix, ' not found in sample info file; please check DNA data prefix')
                                     return(data.table())
                                   }
                                   
                                   data.table(patient_id = sample_info[dna_data_prefix == sample_prefix,
                                                                       patient_id],
                                              syn_coding_variants = data %>%
                                                .[, grepl('synonymous_variant', names(.)), with = F] %>% sum,
                                              non_syn_coding_variants = data %>%
                                                .[, grepl('missense_variant', names(.)), with = F] %>% sum,
                                              stop_gained_variants = data %>%
                                                .[, grepl('stop_gained', names(.)), with = F] %>% sum,
                                              stop_lost_variants = data %>%
                                                .[, grepl('stop_lost', names(.)), with = F] %>% sum,
                                              inframe_indel_variants = data %>%
                                                .[, grepl('inframe_(inser|dele)tion', names(.)), with = F] %>% sum,
                                              frameshift_variants = data %>%
                                                .[, grepl('frameshift_variant', names(.)), with = F] %>% sum)
                                 }))

effects_table = effects_table[naturalorder(effects_table$patient_id)]

sample_info %>% as_tibble() %>%
  mutate_if(grepl(pattern = 'hla_[abc]_[12]', x = names(.)),
            setCellSpec2,
            bg_test = '!grepl(pattern = regexPatterns$allele_exclusion, x = ...)') %>%
  knitr::kable('html', escape = F) %>%
  kable_styling(bootstrap_options = c('striped', 'hover'))

knitr::kable(effects_table) %>%
  kable_styling(bootstrap_options = c('striped', 'hover'))
```

```{r load_prediction_data, message = FALSE}
# parse data
project_predictions = parseEpitopePredictions(path = '3_neolution/predictions_output',
                                              pattern = '_epitopes[.]csv$')

# take subsets based on desired cutoffs
project_predictions_subset = applyCutoffs(predictions = project_predictions,
                                          model = runOptions$neolution$model_cutoff,
                                          expression = runOptions$neolution$expression_cutoff,
                                          selfsim = runOptions$neolution$selfsim_filter)

# exclude alleles
project_predictions_subset = project_predictions_subset[!grepl(pattern = regexPatterns$allele_exclusion,
                                                               x = names(project_predictions_subset))]

# split by xmer
project_predictions_by_xmer = prepareEpitopeLists(list_of_predictions = project_predictions_subset)
```

```{r prepare_order_and_setup_lists}
# bind all together for order list
all_predictions = rbindlist(project_predictions_by_xmer)
all_predictions = all_predictions[naturalorder(patient_id)]
setkey(x = all_predictions, xmer, tumor_peptide_resin)

# take subsets for individual patients for combi-coding setup lists
predictions_by_patient = setNames(object = lapply(sample_info$dna_data_prefix,
                                                  function(prefix) {
                                                    data_subset = all_predictions[sample_prefix == prefix]
                                                    
                                                    data_unique = unique(x = data_subset,
                                                                         by = c('hla_allele', 'tumor_peptide'))
                                                    
                                                    setorder(x = data_unique, xmer, hla_allele, -model_prediction)
                                                    return(data_unique)
                                                  }),
                                  nm = sample_info$patient_id)

# prepare unfiltered predictions to generate some stats
all_predictions_unfiltered = rbindlist(project_predictions)
setnames(x = all_predictions_unfiltered,
         old = grep('affinity|percentile_rank', names(all_predictions_unfiltered), value = T),
         new = c('tumor_affinity_nM', 'tumor_affinity_rank', 'normal_affinity_nM', 'normal_affinity_rank'))
all_predictions_unfiltered_by_patient = setNames(object = lapply(sample_info$patient_id,
                                                                 function(id) {
                                                                   data_subset = all_predictions_unfiltered[patient_id == id]
                                                                   
                                                                   return(data_subset)
                                                                 }),
                                                 nm = sample_info$patient_id)
```

<br>

#### Predicted epitopes summary  
_Yield rates are defined as unique epitopes per non-synonymous coding variant_

```{r generate_prediction_summary, results = 'hide'}
prediction_table = merge(x = data.table(patient_id = names(predictions_by_patient),
                                        total_alleles = sapply(names(all_predictions_unfiltered_by_patient),
                                                         function(id) {
                                                           length(all_predictions_unfiltered_by_patient[[id]][, unique(hla_allele)])
                                                         }),
                                        `total_epitopes_\npredicted` = sapply(all_predictions_unfiltered_by_patient,
                                                                              function(dt) dt[!duplicated(tumor_peptide), .N]),
                                        selected_alleles = sapply(names(predictions_by_patient),
                                                         function(id) {
                                                           length(predictions_by_patient[[id]][, unique(hla_allele)])
                                                         }),
                                        `selected_epitopes_\npredicted` = sapply(predictions_by_patient,
                                                                                 function(dt) dt[!duplicated(tumor_peptide), .N])),
                         y = effects_table[, -c('syn_coding_variants'), with = F],
                         by = 'patient_id')
prediction_table = prediction_table[naturalorder(patient_id)]
prediction_table[, `total_epitope_\nyield_rate_per_allele` := round(x = `total_epitopes_\npredicted` / non_syn_coding_variants / total_alleles,
                                                                    digits = 1)]
prediction_table[, `selected_epitope_\nyield_rate_per_allele` := round(x = `selected_epitopes_\npredicted` / non_syn_coding_variants / selected_alleles,
                                                                       digits = 2)]

order_first = c('patient_id',
                'total_alleles', 'total_epitopes_\npredicted', 'total_epitope_\nyield_rate_per_allele',
                'selected_alleles', 'selected_epitopes_\npredicted', 'selected_epitope_\nyield_rate_per_allele')
setcolorder(x = prediction_table,
            neworder = c(order_first, 
                         setdiff(x = names(prediction_table),
                                 y = order_first)))
```

```{r knit_prediction_summary}
knitr::kable(prediction_table[, -grep('variants', names(prediction_table), value = T), with = F]) %>%
  kable_styling(bootstrap_options = c('striped', 'hover'))
```

<br>

__Total number of unique selected epitopes: `r nrow(unique(all_predictions, by = 'tumor_peptide'))`__  

<br>

#### Predicted epitopes - __All patients__ - master order list (unique peptides)
- __epitopes are sorted by xmer and subsequently by resin__
```{r knit_order_list}
# peptide synth facility list: list all unique epitopes, sorted by xmer and last residue
knitr::kable(all_predictions[, names(all_predictions) %nin%
                               c('sample_prefix', 'hla_allele',
                                 'tumor_affinity_rank', 'model_prediction', 'different_from_self'),
                             with = F] %>% unique(by = 'tumor_peptide')) %>%
  kable_styling(bootstrap_options = c('striped', 'hover'))
```

<br>

#### Predicted epitopes - __Individual patients__ - list for combi-coding setups
- __epitopes predicted to bind to multiple alleles are shown in bold__
- __epitopes with affinity_rank_score >= 20 are marked red to indicate weak HLA binding__
- __epitopes are sorted by xmer first, hla\_allele second, then model_prediction score__
```{r knit_setup_lists, results = 'asis'}
# multimer combi-coding setup list: all epitopes and alleles, sorted by xmer and hla_allele
for (i in 1:length(predictions_by_patient)) {
  cat('\n#####', names(predictions_by_patient)[i], '\n')
  
  # highlight duplicates (epitopes binding to multiple alleles)
  if (any(duplicated(predictions_by_patient[[i]][, tumor_peptide]))) {
    predictions_by_patient[[i]][, -c('sample_prefix', 'tumor_peptide_trunc', 'tumor_peptide_resin'), with = FALSE] %>% as_tibble %>%
      mutate_if(grepl(pattern = 'tumor_affinity_rank', x = names(.)),
                setCellSpec2, bg_test = '... < 20') %>%
      knitr::kable('html', escape = F) %>%
      row_spec(row = predictions_by_patient[[i]][, .I[duplicated(tumor_peptide) | duplicated(tumor_peptide, fromLast = T)]] %>% unique,
               bold = T) %>%
      kable_styling(bootstrap_options = c('striped', 'hover')) %>%
      print()
  } else {
    predictions_by_patient[[i]][, -c('sample_prefix', 'tumor_peptide_trunc', 'tumor_peptide_resin'), with = FALSE] %>% as_tibble %>%
      mutate_if(grepl(pattern = 'tumor_affinity_rank', x = names(.)),
                setCellSpec2, bg_test = '... < 20') %>%
      knitr::kable('html', escape = F) %>%
      kable_styling(bootstrap_options = c('striped', 'hover')) %>%
      print()
  }
}
```

